### PR TITLE
[SCFToCalyx, Calyx, CalyxToHW] Add support for signed div and rem

### PR DESCRIPTION
**Pull request by mortbopet**
Adds support both at the front- and back-end level to ensure that all signed `arith` division/modulo operators work end-to-end.

Notably, this also changes the semantics of the `DivLibPipe` operator in that it no longer provides both a quotient and remainder result. This change aligns the Calyx operators to better fit within an MLIR/CIRCT up/downstream context (`arith` and `comb` operators), but breaks compatibility with the div primitive in the native calyx compiler. A note has been added in the Calyx emitter to indicate of this.

Opted not to add support in StaticLogicToCalyx for the new operations since i assume the existing code duplication will be cleaned up in the near future, with `SCFToCalyx` taking precedence for the shared logic.

### CODE REVIEW

File: lib/Dialect/Calyx/Export/CalyxEmitter.cpp [hunk 1]
```diff
@@ -79,6 +79,15 @@ static Optional<StringRef> unsupportedOpInfo(Operation *op) {
             "compiler (see github.com/cucapra/calyx/issues/1009)";
         return {info};
       })
+      .Case<DivUPipeLibOp, DivSPipeLibOp, RemUPipeLibOp, RemSPipeLibOp>(
+          [](auto) -> Optional<StringRef> {
+            static std::string_view info =
```
Comment (cgyurgyik) [inline lib/Dialect/Calyx/Export/CalyxEmitter.cpp:84]: static constexpr?

File: lib/Dialect/Calyx/Export/CalyxEmitter.cpp [hunk 2]
```diff
@@ -118,11 +127,10 @@ struct ImportTracker {
               return {sCore};
             })
         .Case<SgtLibOp, SltLibOp, SeqLibOp, SneqLibOp, SgeLibOp, SleLibOp,
-              SrshLibOp, MultPipeLibOp, DivPipeLibOp>(
-            [&](auto op) -> FailureOr<StringRef> {
-              static std::string_view sBinaryOperators = "binary_operators";
-              return {sBinaryOperators};
-            })
+              SrshLibOp, MultPipeLibOp>([&](auto op) -> FailureOr<StringRef> {
+          static std::string_view sBinaryOperators = "binary_operators";
```
Comment (cgyurgyik) [inline lib/Dialect/Calyx/Export/CalyxEmitter.cpp:131]: similar here.

**Review by cgyurgyik — COMMENTED** [review]
In general LGTM. However, I think that being able to lowering premium operations such as division/remainder to Calyx native should remain a priority. 

I think it should be possible to have this conversion in the Calyx exporter. If the operation is a division/remainder, lower to the respective Calyx native primitive, e.g., [std_div_pipe](https://github.com/cucapra/calyx/blob/574e4a901a6ac5b0fa8adc115471267b17a0c3e8/primitives/binary_operators.futil#L101-L111). Then, when parsing port names, check to see if the parent is a div/rem operation. If so, carefully choose the correct port names.

An alternative is adding support for separate `div`/`rem` primitives in Calyx native.

File: lib/Conversion/SCFToCalyx/SCFToCalyx.cpp [hunk 3]
```diff
@@ -561,7 +561,8 @@ class BuildOpGroups : public FuncOpPartialLoweringPattern {
                              /// standard arithmetic
                              AddIOp, SubIOp, CmpIOp, ShLIOp, ShRUIOp, ShRSIOp,
                              AndIOp, XOrIOp, OrIOp, ExtUIOp, ExtSIOp, TruncIOp,
-                             MulIOp, DivUIOp, RemUIOp, IndexCastOp>(
+                             MulIOp, DivUIOp, DivSIOp, RemUIOp, RemSIOp,
```
Comment (cgyurgyik) [inline lib/Conversion/SCFToCalyx/SCFToCalyx.cpp:564]: Perhaps add a TODO in the StaticLogicToCalyx so we don't forget this?

**Comment by mortbopet** [comment]
> I think it should be possible to have this conversion in the Calyx exporter.

I'll have a look at the implementation which you suggested. I agree that it remains a priority to support interchange in the transition period, but the CIRCT flow should nevertheless take precedence when additions make headway towards end-to-end-in-CIRCT. Which, i think and hope, is a shared goal for the people actively contributing to the CIRCT Calyx dialect.

**Comment by cgyurgyik** [comment]
> > I think it should be possible to have this conversion in the Calyx exporter.
> 
> I'll have a look at the implementation which you suggested. I agree that it remains a priority to support interchange in the transition period, but the CIRCT flow should nevertheless take precedence when additions make headway towards end-to-end-in-CIRCT. Which, i think and hope, is a shared goal for the people actively contributing to the CIRCT Calyx dialect.

Agreed! I just don't foresee a Calyx-complete compiler in CIRCT in the near future, and would hate to have this block others :-)

**Comment by mortbopet** [comment]
Sometimes a bit of blockage can also serve as an incentive to focus efforts into fixing said blockage :thinking: ... :wink:.

**Comment by mortbopet** [comment]
Emitter support added in [6e920a3](https://github.com/llvm/circt/pull/3238/commits/6e920a3014fffe6ac65978bf122b3871efcecf47).

File: lib/Dialect/Calyx/Export/CalyxEmitter.cpp [hunk 4]
```diff
@@ -545,8 +541,12 @@ void Emitter::emitComponent(ComponentOp op) {
                 SubLibOp, ShruLibOp, RshLibOp, SrshLibOp, LshLibOp, AndLibOp,
                 NotLibOp, OrLibOp, XorLibOp, WireLibOp>(
               [&](auto op) { emitLibraryPrimTypedByFirstInputPort(op); })
-          .Case<MultPipeLibOp, DivUPipeLibOp, DivSPipeLibOp>(
+          .Case<MultPipeLibOp>(
               [&](auto op) { emitLibraryPrimTypedByFirstOutputPort(op); })
+          .Case<RemUPipeLibOp, RemSPipeLibOp, DivUPipeLibOp, DivSPipeLibOp>(
```
Comment (cgyurgyik) [inline lib/Dialect/Calyx/Export/CalyxEmitter.cpp:546]: Signed remainder/divpipe should be `std_sdiv_pipe`

Ref: https://github.com/cucapra/calyx/blob/574e4a901a6ac5b0fa8adc115471267b17a0c3e8/primitives/binary_operators.futil#L128

File: lib/Dialect/Calyx/Export/CalyxEmitter.cpp [hunk 5]
```diff
@@ -545,8 +541,12 @@ void Emitter::emitComponent(ComponentOp op) {
...
+              [&](auto op) {
+                emitLibraryPrimTypedByFirstOutputPort(op, {"std_div_pipe"});
```
Comment (cgyurgyik) [inline lib/Dialect/Calyx/Export/CalyxEmitter.cpp:548]: Also perhaps `"std_div_pipe"` (no brackets) isn't sufficient here? Also perhaps `/*calyxLibName=*/`

**Review by cgyurgyik — APPROVED** [review]
Thanks for the native compatibility changes :-)

File: lib/Dialect/Calyx/Export/CalyxEmitter.cpp (hunk 5 above)
Comment (mortbopet) [reply lib/Dialect/Calyx/Export/CalyxEmitter.cpp:548]: `const char*` to `Optional<StringRef>` seems to a be one step too much for the compiler to deduce, hence the brackets.

**Pull request merged**